commands/update_dependencies.go

```diff
@@ -51,43 +55,87 @@ func updateDependenciesRun(flags updateDependenciesFlags) error {
+		allDependencies := append(config.Metadata.Dependencies, newVersions...)
+
+		for _, constraint := range config.Metadata.DependencyConstraints {
+			// Manually lookup the existent dependency name from the buildpack.toml since this
+			// isn't specified via the dep-server
+			dependencyName := internal.FindDependencyName(constraint.ID, config)
+
+			// Filter allDependencies for only those that match the constraint
+			// mds is a just the right number of deps for the constraint
+			mds, err := internal.GetCargoDependenciesWithinConstraint(allDependencies, constraint, dependencyName)
 			if err != nil {
 				return err
 			}
-			allDependencies[constraint.ID] = dependencies
+
+			matchingDependencies = append(matchingDependencies, mds...)
+			if len(matchingDependencies) > 0 {
+				config.Metadata.Dependencies = matchingDependencies
+			}
 		}
+	} else {
+		// All internal.Dependencies from the dep-server
+		allDependencies := map[string][]internal.Dependency{}
+		// All cargo.ConfigMetadataDependencies that match one of the given constraints
+		var matchingDependencies []cargo.ConfigMetadataDependency
+
+		for _, constraint := range config.Metadata.DependencyConstraints {
+			// Only query the API once per unique dependency
+			dependencies, ok := allDependencies[constraint.ID]
+			if !ok {
+				var err error
+				dependencies, err = internal.GetAllDependencies(api, constraint.ID)
+				if err != nil {
+					return err
+				}
+				allDependencies[constraint.ID] = dependencies
+			}
 
-		// Manually lookup the existent dependency name from the buildpack.toml since this
-		// isn't specified via the dep-server
-		dependencyName := internal.FindDependencyName(constraint.ID, config)
+			// Manually lookup the existent dependency name from the buildpack.toml since this
+			// isn't specified via the dep-server
+			dependencyName := internal.FindDependencyName(constraint.ID, config)
 
-		// Filter allDependencies for only those that match the constraint
-		mds, err := internal.GetDependenciesWithinConstraint(dependencies, constraint, dependencyName)
-		if err != nil {
-			return err
+			// Filter allDependencies for only those that match the constraint
+			// mds is a just the right number of deps for the constraint
+			mds, err := internal.GetDependenciesWithinConstraint(dependencies, constraint, dependencyName)
+			if err != nil {
+				return err
+			}
+			matchingDependencies = append(matchingDependencies, mds...)
 		}
-		matchingDependencies = append(matchingDependencies, mds...)
-	}
 
-	if len(matchingDependencies) > 0 {
-		config.Metadata.Dependencies = matchingDependencies
+		if len(matchingDependencies) > 0 {
+			config.Metadata.Dependencies = matchingDependencies
+		}
 	}
 
-	newVersions := []string{}
+	newVersions := map[string]string{}
 	for _, d := range config.Metadata.Dependencies {
 		if _, ok := originalVersions[d.Version]; !ok {
-			newVersions = append(newVersions, d.Version)
+			newVersions[d.Version] = ""
 		}
```

Comment:
This change is here because in the `metadata-file` path, there may be version duplicates (that differ by stack). Rhe `newVersions` list is just used to print out a list of the new versions added, so we don't want duplicates.